### PR TITLE
Verify section ids

### PR DESCRIPTION
**Pull request by pledbrook**
Adds duplicate checks of section IDs in the user guide to the build and also checks if any sections have no IDs. Section IDs have been inserted/modified to satisfy these new criteria.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.gradle.api.tasks.TaskAction
+import org.w3c.dom.NodeList
+
+import javax.xml.parsers.DocumentBuilderFactory
+import javax.xml.xpath.XPathConstants
+import javax.xml.xpath.XPathFactory
+
+
+/**
+ * Created by pledbrook on 03/05/2016.
```
Comment (bmuschko) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: Can we provide a more expressive description here?

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ */
+@CompileStatic
+class UserGuideSectionVerifierTask extends DefaultTask {
```
Comment (bmuschko) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: I'd love to see a test for the custom task e.g. implemented via `ProjectBuilder`. The test would create a new instance of a project, create a task of this type and then execute the action by calling `execute()`. The test should cover all possible use cases.

Comment (bmuschko) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: I tend to not use the "Task" suffix to the class name as it's usually implied when setting the type of an enhance tasks.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        for (int i in 0..<sectionIds.length) {
+            idList << sectionIds.item(i).nodeValue
+        }
+
+        return new ValidationResult(
+                missingSectionIdCount: sectionsWithoutIds.length,
+                sectionIds: idList)
+    }
+
+    private String generateExceptionMessage(
```
Comment (bmuschko) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: Can we break up the method into two methods? One of duplicate IDs and one for missing IDs.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                allResults.inject(0) { int sum, String filename, ValidationResult result ->
+            sum + result.missingSectionIdCount
+        }
+
+        if (duplicateSectionIds || allResults.any { filename, result -> result.missingSectionIdCount }) {
+            throw new GradleException(generateExceptionMessage(duplicateSectionIds, allResults))
+        }
+    }
+
+    public ValidationResult validateSections(File file) {
```
Comment (bmuschko) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: Should use the `private` modifier.

File: subprojects/docs/docs.gradle [hunk 5]
```diff
@@ -266,8 +267,12 @@ task dslHtml(type: Docbook2Xhtml) {
     ext.entryPoint = "$destDir/index.html"
 }
 
+task checkSectionIds(type: UserGuideSectionVerifierTask) {
+    docbookFiles = fileTree(userguideSrcDir) { include "**/*.xml" }
+}
+
 // This is used in the distribution and for the online version
-task userguideDocbook(type: UserGuideTransformTask, dependsOn: [samples, samplesDocbook, defaultImports]) {
+task userguideDocbook(type: UserGuideTransformTask, dependsOn: [checkSectionIds, samples, samplesDocbook, defaultImports]) {
```
Comment (bmuschko) [inline subprojects/docs/docs.gradle]: We'll likely want to check IDs as the first thing in the list of task dependencies so we can fail early. We could use a `mustRunAfter` statement to define the task execution order.

**Comment by bmuschko** [comment]
Would you mind rebasing your changes?

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy (hunk 3 above)
Comment (pledbrook) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: We could, but I'm not sure of the value in this case. We still need a method to generate the overall exception message.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy (hunk 2 above)
Comment (pledbrook) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: On the other hand, all the other tasks imported by _docs.gradle_ have a `Task` suffix :smile:

File: subprojects/docs/docs.gradle (hunk 5 above)
Comment (pledbrook) [inline subprojects/docs/docs.gradle]: I'm not sure of the best way to do this. The only reliable way in my mind is to apply `mustRunAfter(checkSectionIds)` to all the other tasks that `userguideHtml` depends on.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy (hunk 2 above)
Comment (bmuschko) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: Some do have the `Task` prefix. We should unify that at some point of time. I think we should stick with the general convention we have in Gradle core. All public task types do not add the `Task` suffix e.g. `Copy`, `JavaCompile` etc.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy (hunk 3 above)
Comment (bmuschko) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: The value would be better readability and maintainability in the long run. Try to see if it makes sense. If it complicates the code then leave it as is.

File: subprojects/docs/docs.gradle (hunk 5 above)
Comment (bmuschko) [inline subprojects/docs/docs.gradle]: Correct. I was think along the lines of this:

    [samples, samplesDocbook, defaultImports]*.mustRunAfter(checkSectionIds)

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy (hunk 2 above)
Comment (pledbrook) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifierTask.groovy]: I have everything else done, but the test case is probably going to take a bit longer.

**Comment by pledbrook** [comment]
@bmuschko Everything done I think. Hopefully there won't be conflicts when new doc changes are added, but there probably will be.

**Comment by bmuschko** [comment]
@pledbrook Looks like there's still a merge conflict on `subprojects/docs/src/docs/userguide/gradleDaemon.xml`. Can you have a look at one? If you don't get to it, I'll try to resolve the conflict myself and merge.

**Comment by pledbrook** [comment]
Try again, but please do so as soon as possible. This commit touches so many user guide source files that there's a good chance that any small change to the user guide in the mean time will cause a conflict.

**Comment by bmuschko** [comment]
@pledbrook Thanks for updating. I just merged the PR.

**Pull request merged**